Redesigned logger api (#2)
anz-bank/pkg: log/examples.md

        "stuff": 1,
    })

    // Any logs at this point will have context level fields
    contextLevelField(ctx)
    logLevelField(ctx)
    logGlobalField(ctx)
}

func contextLevelField(ctx context.Context) {
    // This is expected to log something like
    // 2019-12-12T08:23:59.210878+11:00 just=stuff stuff=1 DEBUG Debug
    log.From(ctx).Warn("Warn")
}

func logLevelField(ctx context.Context) {

>     // Log level fields are fields that are not stored into the context logger

Reviewer: ```suggestion
    // Log level fields are fields that are not stored in the context logger fields
```

    log.From(ctx).Debug("Debug")

    ctx = context.WithValue(ctx, 1, 2)
    ctx = context.WithValue(ctx, "random", "random")
    ctx = context.WithValue(ctx, "foo", "bar")

    // this will use all the values registered in context and it will also detect changes
    // of value from context, it will not log fields whose keys are not registered through the
    // WithGlobalFields API
    // This is expected to log something like
    // 2019-12-12T08:23:59.210878+11:00 random=random 1=2 DEBUG Debug
    log.From(ctx).Debug("Debug")

    // it is also possible to mix this with other types of fields
    // all of the fields will be combined
    ctx = log.WithFields(ctx, log.MultipleFields{42:43})

    // This is expected to log something like
    // 2019-12-12T08:23:59.210878+11:00 random=random 1=2 42=43 golang=very cool DEBUG Debug
    log.From(ctx, log.NewField("golang", "very cool")).Debug("Debug")
}
```

Due to the three sources of fields, it is important to understand the each source has a different priority when there are overlapping keys in the fields since all keys in the fields must be unique.

```go
func logPrecedenceDemo(ctx context.Context) {

>     // Each of field types has a priority level with global being the lowest, context level in the middle,

Reviewer: ```suggestion
    // Each field type has a priority level with global being the lowest, context level in the middle,
```

    // and log specific the highest.
    // The priority levels determine how which fields replace which since keys in fields must be unique. So
    // this means that context level fields will replace global level fields and log specific fields will
    // replace all fields

    // register keys
    ctx = log.WithGlobalFields(ctx, "foo", "bar", 42)
    ctx = context.WithValue(ctx, "foo", "foo")
    ctx = context.WithValue(ctx, "bar", "bar")
    ctx = context.WithValue(ctx, 42, 42)

>     // logging at this point, will only log global level fields since there are not context level fields

Reviewer: ```suggestion
    // Logging at this point, will only log global level fields since there are not context level fields
```

    // registered to the logger
    // This is expected to log something like
    // 2019-12-12T08:23:59.210878+11:00 foo=foo bar=bar 42=42 DEBUG Debug
    log.From(ctx).Debug("Debug")

>     // add context level fields

Reviewer: ```suggestion
    // Add context level fields
```

    ctx = log.WithFields(ctx, log.MultipleFields{"foo": "fooooo", "hello": "there"})
    ctx = context.WithValue(
            ctx, "foo", "change the value to show the order doesn't matter, context level fields still win")

>     // logging at this point use the corresponding value of "foo" from the context level fields

Reviewer: ```suggestion
    // Logging at this point use the corresponding value of "foo" from the context level fields
```

    // but the rest of the global level fields will always be logged
    // non overlapping fields will always be added
    // This is expected to log something like
    // 2019-12-12T08:23:59.210878+11:00 foo=fooooo bar=bar 42=42 hello=there DEBUG Debug
    log.From(ctx).Debug("Debug")

>     // logging with log specific fields will replace all overlapping fields

Reviewer: ```suggestion
    // Logging with log specific fields will replace all overlapping fields
```
